### PR TITLE
Refs [TASK][35928] Design Database

### PR DESCRIPTION
**Pull request by huydd-2678**
## Related Tickets
- [#Task 35928: Design Database](https://edu-redmine.sun-asterisk.vn/issues/35928)
## WHAT
- Create room database, dao, entity
- Apply Koin
- Database local design:
Table Product
+ Id
+ Name
+ Image
+ Rating
+ Category
+ CategoryId
+ Stock
+ Materail
+ Sold
+ Color
+ isFavorite
+ isAddedToCart
+ quantityInCart
Table Bill
+ Id
+ Time
+ TotalItem
+ TotalPrice
+ ProductName
+ ProductPrice
+ ProductQuantity
+ ProductImage
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sun/dummyshop/data/repository/BillRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import com.sun.dummyshop.data.source.BillDataSource
+import io.reactivex.rxjava3.core.Completable
+import io.reactivex.rxjava3.core.Observable
+
+class BillRepositoryImpl(
+    private val local: BillDataSource.Local
+) : BillRepository {
+
+    override fun getBills(): Observable<List<Bill>> =
+        local.getBills()
```
Comment (thonc-2135) [inline app/src/main/java/com/sun/dummyshop/data/repository/BillRepositoryImpl.kt]: sao k để cùng 1 dòng

File: app/src/main/java/com/sun/dummyshop/data/repository/ProductRepository.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.sun.dummyshop.data.repository
+
+import com.sun.dummyshop.data.model.Product
+import io.reactivex.rxjava3.core.Completable
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.core.Single
+
+interface ProductRepository {
+    fun getFavoriteProducts(): Observable<List<Product>>
+    fun insertFavoriteProduct(product: Product): Completable
```
Comment (thonc-2135) [inline app/src/main/java/com/sun/dummyshop/data/repository/ProductRepository.kt:10]: chỉ trả về mỗi completable thôi à

File: app/src/main/java/com/sun/dummyshop/data/source/local/dao/ProductDAO.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package com.sun.dummyshop.data.source.local.dao
+
+import androidx.room.*
+import com.sun.dummyshop.data.model.Product
+import io.reactivex.rxjava3.core.Completable
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.core.Single
+
+@Dao
+interface ProductDAO {
```
Comment (thonc-2135) [inline app/src/main/java/com/sun/dummyshop/data/source/local/dao/ProductDAO.kt]:
```suggestion
interface ProductDao {
```

File: app/src/main/java/com/sun/dummyshop/data/source/local/dao/ProductDAO.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    fun insertFavoriteProduct(product: Product): Completable
+
+    @Delete
+    fun deleteFavoriteProduct(product: Product): Completable
+
+    @Update
+    fun updateFavoriteProduct(product: Product): Completable
+
+    @Query("SELECT * FROM product WHERE id = :id AND isFavorite = 1")
+    fun isFavoriteProduct(id: String): Single<List<Product>>
```
Comment (thonc-2135) [inline app/src/main/java/com/sun/dummyshop/data/source/local/dao/ProductDAO.kt]: hàm là isXYZ mà lại trả về list, nên đổi tên

File: app/src/main/java/com/sun/dummyshop/data/model/Bill.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import androidx.room.ColumnInfo
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import com.google.gson.annotations.SerializedName
+
+@Entity(tableName = "bill")
+data class Bill(
+    @PrimaryKey
+    @SerializedName("date")
+    @ColumnInfo
```
Comment (huytq-2422) [inline app/src/main/java/com/sun/dummyshop/data/model/Bill.kt]:
```suggestion
    @ColumnInfo(name = "date")
```

File: app/src/main/java/com/sun/dummyshop/DummyShopApplication.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import org.koin.core.context.startKoin
+
+class DummyShopApplication : Application() {
+
+    override fun onCreate() {
+        super.onCreate()
+        startKoin {
+            androidContext(this@DummyShopApplication)
+            modules(
+                listOf(
```
Comment (huytq-2422) [inline app/src/main/java/com/sun/dummyshop/DummyShopApplication.kt]: hình như truyền vararg được, không cần bỏ vào list 
```suggestion
            modules(
```

File: app/src/main/java/com/sun/dummyshop/data/repository/ProductRepository.kt (hunk 2 above)
Comment (huydd-2678) [reply app/src/main/java/com/sun/dummyshop/data/repository/ProductRepository.kt:10]: vâng, tại chỉ cần biết là insert được hay không thôi ạ

File: app/src/main/java/com/sun/dummyshop/DummyShopApplication.kt (hunk 6 above)
Comment (huydd-2678) [reply app/src/main/java/com/sun/dummyshop/DummyShopApplication.kt]: vâng

File: app/src/main/java/com/sun/dummyshop/data/source/local/dao/ProductDAO.kt (hunk 4 above)
Comment (huydd-2678) [reply app/src/main/java/com/sun/dummyshop/data/source/local/dao/ProductDAO.kt]: vâng

File: app/src/main/java/com/sun/dummyshop/di/RepositoryModule.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import com.sun.dummyshop.data.source.local.BillLocalDataSource
+import com.sun.dummyshop.data.source.local.ProductLocalDataSource
+import org.koin.dsl.module
+
+val productRepositoryModule = module {
+    single<ProductDataSource.Local> { ProductLocalDataSource(get()) }
+    single<ProductRepository> { ProductRepositoryImpl(get()) }
+}
+
+val billRepositoryModule = module {
```
Comment (datdt-1935) [inline app/src/main/java/com/sun/dummyshop/di/RepositoryModule.kt]: gộp chung lại thành `repositoryModule` cho gọn nhé

Comment (huydd-2678) [reply app/src/main/java/com/sun/dummyshop/di/RepositoryModule.kt]: vâng

**Pull request merged**